framework/src/st_things/things_stack/utils/things_network.c

```diff
@@ -318,13 +358,94 @@ void things_wifi_soft_ap_sta_left(void)
 	THINGS_LOG_V(TAG, "T%d --> %s", getpid(), __FUNCTION__);
 }
 
+void things_wifi_list_cleanup(void)
+{
+	if (g_wifi_scan_info != NULL) {
+		// free old wifi list
+
+		access_point_info_s* piter;
+		access_point_info_s* pdel;
+
+		piter = pdel = g_wifi_scan_info;
+
+		while (piter != NULL) {
+			pdel = piter;
+			piter = piter->next;
+
+			things_free(pdel);
+		}
+
+		g_wifi_scan_info = NULL;
+	}
+
+	g_wifi_count = 0;
+}
+
 void things_wifi_scan_done(wifi_manager_scan_info_s **scan_result, int res)
 {
 	THINGS_LOG_V(TAG, "T%d --> %s", getpid(), __FUNCTION__);
 	/* Make sure you copy the scan results onto a local data structure.
 	 * It will be deleted soon eventually as you exit this function.
 	 */
-//TO DO
+	if (scan_result == NULL) {
+		return;
+	}
+
+	pthread_mutex_lock(&g_ap_list_mutex);
+
+	things_wifi_list_cleanup();
+	wifi_manager_scan_info_s *wifi_scan_iter = *scan_result;
+	access_point_info_s *pinfo = NULL;
+	access_point_info_s *p_last_info = NULL;
+
+	while (wifi_scan_iter != NULL) {
+		if (wifi_scan_iter->ssid != NULL) {
+			pinfo = (access_point_info_s*)things_malloc(sizeof(access_point_info_s));
+			pinfo->next = NULL;
+			snprintf(pinfo->e_ssid, WIFIMGR_SSID_LEN, "%s", wifi_scan_iter->ssid);
+			snprintf(pinfo->bss_id, WIFIMGR_MACADDR_STR_LEN, "%s", wifi_scan_iter->bssid);
+			snprintf(pinfo->signal_level, MAX_LEVEL_SIGNAL, "%d", wifi_scan_iter->rssi);
+			//strncpy(pinfo->security_key, wifi_scan_iter->ssid, MAX_SECUIRTYKEY_LEN);
+			//strncpy(pinfo->auth_type, wifi_scan_iter->ssid, MAX_TYPE_AUTH);
+			//strncpy(pinfo->enc_type, wifi_scan_iter->ssid, MAX_TYPE_ENC);
+
+			if (g_wifi_scan_info == NULL) {
+				g_wifi_scan_info = pinfo;
+			} else {
+				p_last_info->next = pinfo;
+			}
+
+			p_last_info = pinfo;
+			g_wifi_count++;
+		}
+
+		wifi_scan_iter = wifi_scan_iter->next;
+	}
+
+	pinfo = g_wifi_scan_info;
+	while (pinfo != NULL) {
+		THINGS_LOG_V(TAG, "WiFi AP - SSID: %-20s, WiFi AP BSSID: %-20s, WiFi Rssi: %d\n",
+				   pinfo->e_ssid, pinfo->bss_id, pinfo->signal_level);
+
+		pinfo = pinfo->next;
+	}
+
+	THINGS_LOG_V(TAG, "WiFi List Count is (%d)", g_wifi_count);
+	pthread_mutex_unlock(&g_ap_list_mutex);
+	return;
+}
+
+int things_wifi_scan_ap(void)
+{
+	while (dm_is_es_complete() == false) {
```

Comment:
in this case, you couldn't get out of the loop until es_complete.

In the following statement, you may have wanted to use Sleep, but sleep is not applied here
[things_stack.c] line 418: static void *__attribute__((optimize("O0"))) auto_scanning_loop(void)

Reply:
fixed.